# Emerging Technology Tasks
Solutions to my tasks assessment for emerging technology module. Author: Nathan Garrihy

### Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places.
I used a googol value to store the numbers on the right side of the decimal place as a whole number to allow 100 decimal place output to screen. The program iterates a few guesses to what the sqrt is and then returns the square root of 2 value when the guess is equal to the previous guess!

[1] Newton Raphson's Method - https://surajregmi.medium.com/how-to-calculate-the-square-root-of-a-number-newton-raphson-method-f8007714f64

[2] NASA's Square Root of 2 to 1 million decimals - https://apod.nasa.gov/htmltest/gifcity/sqrt2.1mil

![NewtonRaphson](https://media.geeksforgeeks.org/wp-content/cdn-uploads/newtonraphsonformula.png "Newton's Method")


In [164]:
def Sqrt2():
    # This constant is not fully required but can be changed
    # in order to find the square root of a different number
    SQRTNUM = 2;
    # googol = 10 to the power of 100
    GOOGOL = 10**100
    # 'num' variable  = num value of 2 * googol^2. This
    # moves numbers to the left side of the decimal place
    num = SQRTNUM *(GOOGOL)**2
    a = 0
    b = GOOGOL

    # Loop until accuracy is sufficiant
    # Keeps running until a == b
    while a != b:
        # Set a to b
        a = b
        # Get the floor division + Bit shift num 1 place to the right
        b = (a+(num//a)) >> 1
    # finalVal contains the numbers to the right of decimal place
    finalVal = b-(GOOGOL)
    # Code for potential future re-use could
    # change SQRTNUM and return leftNum and finalVaL
    # leftNum = int((b - finalVal)/GOOGOL)
    
    return "1.%d"%finalVal

Print square root of 2

In [165]:
print(Sqrt2())

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


ReadME file is located on the main branch of my github repository

### Using scipy.stats to verify this value  of a given table and calculate the associated p value
Here, I perform a Chi-square test of independence of variables in a contingency table. This tests the values stored in whiteCol, blueCol and noCol and returns values which can be used to determine whether or not the categorical variables
are independent. 

[1] scipy.stats documentation - https://docs.scipy.org/doc/scipy/reference/stats.html

[2] scipy.stats.chi2contingency documentation - https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html 

In [5]:
# scipy.stats module for probability distributions and 
import scipy.stats as stats

# Table row values stored as arrays
whiteCol = [90, 60, 104, 95]
blueCol = [30, 50, 51, 20]
noCol = [30, 40, 45, 35]

# Values put into 2-Dimentional Array
a = [whiteCol]+[blueCol]+[noCol]

# Use stats.chi2_contingency() function to compute the chi-square 
# statistic and p-value for the test of the observed frequencies 
# in the contingency table. The degrees of freedom which is the number 
# of independent observations in a set of data. It also returns the 
# expected frequencies, based on the marginal sums of the table.
V, p, dof, expected = stats.chi2_contingency(a)

Print stats.chi2_contingency() values to screen

In [14]:
print("========================================================\n",
      "Chi-squared value based on the table =",V,"\n",
      "The p-value of the test =", p,"\n",
      "Degrees of freedom =", dof,"\n",
      "The expected frequencies:\n", expected,"\n"
     "========================================================")

 Chi-squared value based on the table = 24.5712028585826 
 The p-value of the test = 0.0004098425861096696 
 Degrees of freedom = 6 
 The expected frequencies:
 [[ 80.53846154  80.53846154 107.38461538  80.53846154]
 [ 34.84615385  34.84615385  46.46153846  34.84615385]
 [ 34.61538462  34.61538462  46.15384615  34.61538462]] 


![ChiSquaredDistribution](https://www.statisticshowto.com/wp-content/uploads/2014/11/chi-square-distribution.png "Chi Squared Distribution")
##### Chi Squared Distribution
In this graph the y-axis represents the p value and the x-axis represents the value for x squared. Each coloured line represents a different value for degrees of freedom.

- The [Wikipedia article](https://en.wikipedia.org/w/index.php?title=Chi-squared_test&oldid=983024096) stated that the approximate chi-squared value based on the table was 24.6, which was correct as my result was 24.57, which rounds up to 24.6.
- The 'p' value was <0.001 which is lower than the significance level (0.05).
- The 'degrees of freedom' = '(r-1)(c-1)' where r = total number of rows and c = total number of columns. In this case the degrees of freedom is 6, so there are 6 independant values that the statistical analysis can estimate.
- The expected frequencies are estimated frequencies for each cell based on our values.

**The null hypothesis is that each person's neighbourhood of residence is independent of the person's occupational classification. I used this chi square test to check if a relationship exists between the categories and based on my results, I reject this null hypothesis. There is no clear evidence to make me accept this hypothesis as a corrolation between neightborhood of residence and occupational classifaction can be seen, proving lack of independence**